File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/helpers/SedMapperFactory.java
package no.nav.melosys.eessi.service.sed.helpers;

import java.util.Map;

import com.google.common.collect.ImmutableMap;
import com.google.common.collect.Maps;
import lombok.experimental.UtilityClass;
import no.nav.melosys.eessi.models.SedType;
import no.nav.melosys.eessi.models.exception.MappingException;
import no.nav.melosys.eessi.service.sed.mapper.SedMapper;
import no.nav.melosys.eessi.service.sed.mapper.horisontal.HorisontalSedMapper;
import no.nav.melosys.eessi.service.sed.mapper.lovvalg.*;

@UtilityClass
public class SedMapperFactory {

    static final Map<SedType, SedMapper> SED_MAPPERS =
            Maps.immutableEnumMap(ImmutableMap.<SedType, SedMapper>builder()
                    .put(SedType.A001, new A001Mapper())
                    .put(SedType.A002, new A002Mapper())
                    .put(SedType.A003, new A003Mapper())
                    .put(SedType.A005, new A005Mapper())
                    .put(SedType.A008, new A008Mapper())
                    .put(SedType.A009, new A009Mapper())
                    .put(SedType.A010, new A010Mapper())
                    .put(SedType.A011, new A011Mapper())
                    .put(SedType.A012, new A012Mapper())

                    .put(SedType.H001, new HorisontalSedMapper(SedType.H001))
                    .put(SedType.H003, new HorisontalSedMapper(SedType.H003))
                    .put(SedType.H004, new HorisontalSedMapper(SedType.H004))
                    .put(SedType.H005, new HorisontalSedMapper(SedType.H005))
                    .put(SedType.H010, new HorisontalSedMapper(SedType.H010))
                    .put(SedType.H011, new HorisontalSedMapper(SedType.H011))
                    .put(SedType.H020, new HorisontalSedMapper(SedType.H020))
                    .put(SedType.H061, new HorisontalSedMapper(SedType.H061))
                    .put(SedType.H065, new HorisontalSedMapper(SedType.H065))
                    .put(SedType.H070, new HorisontalSedMapper(SedType.H070))
                    .put(SedType.H120, new HorisontalSedMapper(SedType.H120))
                    .put(SedType.H121, new HorisontalSedMapper(SedType.H121))
                    .put(SedType.H130, new HorisontalSedMapper(SedType.H130))

                    .build());

    public static SedMapper sedMapper(SedType sedType) throws MappingException {
        if (!SED_MAPPERS.containsKey(sedType)) {
            throw new MappingException("Sed-type " + sedType.name() + " støttes ikke");
        }
        return SED_MAPPERS.get(sedType);
    }
}

File: melosys-eessi-app/src/main/java/no/nav/melosys/eessi/service/sed/mapper/SedMapperFactory.kt
package no.nav.melosys.eessi.service.sed.mapper

import io.getunleash.Unleash
import mu.KotlinLogging
import no.nav.melosys.eessi.config.featuretoggle.ToggleName.CDM_4_4
import no.nav.melosys.eessi.controller.dto.SedDataDto
import no.nav.melosys.eessi.models.SedType
import no.nav.melosys.eessi.models.exception.MappingException
import no.nav.melosys.eessi.models.sed.Konstanter.DEFAULT_SED_G_VER
import no.nav.melosys.eessi.models.sed.Konstanter.SED_VER_CDM_4_4
import no.nav.melosys.eessi.models.sed.SED
import no.nav.melosys.eessi.service.sed.LandkodeMapper
import no.nav.melosys.eessi.service.sed.mapper.til_sed.SedMapper
import no.nav.melosys.eessi.service.sed.mapper.til_sed.administrativ.X008Mapper
import no.nav.melosys.eessi.service.sed.mapper.til_sed.horisontal.HorisontalSedMapper
import no.nav.melosys.eessi.service.sed.mapper.til_sed.lovvalg.*
import org.springframework.stereotype.Component

private val log = KotlinLogging.logger { }

@Component
class SedMapperFactory(private val unleash: Unleash) {

    private val sedMappers: Map<SedType, SedMapper> = mapOf(
        SedType.A001 to A001Mapper(),
        SedType.A002 to A002Mapper(),
        SedType.A003 to A003Mapper(),
        SedType.A004 to A004Mapper(),
        SedType.A005 to A005Mapper(),
        SedType.A008 to A008Mapper(unleash),
        SedType.A009 to A009Mapper(),
        SedType.A010 to A010Mapper(),
        SedType.A011 to A011Mapper(),
        SedType.A012 to A012Mapper(),
        SedType.X008 to X008Mapper(),
        SedType.H001 to HorisontalSedMapper(SedType.H001),
        SedType.H003 to HorisontalSedMapper(SedType.H003),
        SedType.H004 to HorisontalSedMapper(SedType.H004),
        SedType.H005 to HorisontalSedMapper(SedType.H005),
        SedType.H010 to HorisontalSedMapper(SedType.H010),
        SedType.H011 to HorisontalSedMapper(SedType.H011),
        SedType.H020 to HorisontalSedMapper(SedType.H020),
        SedType.H021 to HorisontalSedMapper(SedType.H021),
        SedType.H061 to HorisontalSedMapper(SedType.H061),
        SedType.H062 to HorisontalSedMapper(SedType.H062),
        SedType.H065 to HorisontalSedMapper(SedType.H065),
        SedType.H066 to HorisontalSedMapper(SedType.H066),
        SedType.H070 to HorisontalSedMapper(SedType.H070),
        SedType.H120 to HorisontalSedMapper(SedType.H120),
        SedType.H121 to HorisontalSedMapper(SedType.H121),
        SedType.H130 to HorisontalSedMapper(SedType.H130)
    )

    fun sedMapper(sedType: SedType): SedMapper {
        return sedMappers[sedType]
            ?: throw MappingException("Sed-type ${sedType.name} støttes ikke")
    }

    fun mapTilSed(sedType: SedType, sedDataDto: SedDataDto): SED {
        val sed = sedMapper(sedType).mapTilSed(sedDataDto)
        if (unleash.isEnabled(CDM_4_4)) {
            sed.sedGVer = DEFAULT_SED_G_VER
            sed.sedVer = SED_VER_CDM_4_4
            return sed
        }
        return konverterKosovoTilUkjent(sed, sedDataDto.gsakSaksnummer)
    }

    private fun konverterKosovoTilUkjent(sed: SED, gsakSaksnummer: Long?): SED {
        val person = sed.finnPerson().orElse(null) ?: return sed

        person.statsborgerskap = person.statsborgerskap.map {
            if (it?.land == LandkodeMapper.KOSOVO_LANDKODE_ISO2) {
                log.info("Endrer statsborgerskap fra Kosovo til Ukjent. gsakSaksnummer: $gsakSaksnummer")
                it.copy(land = LandkodeMapper.UKJENT_LANDKODE_ISO2)
            } else it
        }
        return sed
    }
}
